frame/support/src/traits/hooks.rs

```diff
@@ -135,21 +136,32 @@ pub trait OnRuntimeUpgrade {
 
 	/// Execute some pre-checks prior to a runtime upgrade.
 	///
+	/// Return an encoded state digest which will be passed to `post_upgrade` after upgrading for
+	/// post-check.
+	///
 	/// This hook is never meant to be executed on-chain but is meant to be used by testing tools.
 	#[cfg(feature = "try-runtime")]
-	fn pre_upgrade() -> Result<(), &'static str> {
-		Ok(())
+	fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+		Ok(Vec::new())
 	}
 
 	/// Execute some post-checks after a runtime upgrade.
 	///
+	/// The encoded state digest generated by `pre_upgrade` before upgrading will be passed in for
+	/// post-check. NOTE: if `pre_upgrade` is not implemented an empty vec will be passed in, in
+	/// such case `post_upgrade` should ignore it.
+	///
 	/// This hook is never meant to be executed on-chain but is meant to be used by testing tools.
 	#[cfg(feature = "try-runtime")]
-	fn post_upgrade() -> Result<(), &'static str> {
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
 		Ok(())
 	}
 }
 
+// A helper struct to help encode/decode tuple
+#[derive(Default, Encode, Decode)]
+struct TupleCodecHelper(Vec<Vec<u8>>);
```

Comment:
```suggestion
type TupleCodecHelper = Vec<Vec<u8>>;
```

Also works, no?

Reply:
Or, you can just name `Vec<Vec<u8>>` everywhere.

Reply:
Type alias may not work because we can't implement encode/decode traits for it, `Vec<Vec<u8>>`is a good idea, I had changed to it.